File: day03/part1.py
_TESTCASE = """\
00100
11110
10110
10111
10101
01111
00111
11100
10000
11001
00010
01010
""".splitlines()


def compute(data):
    """
    >>> compute(_TESTCASE)
    198
    """
    gamma = ""
    epsilon = ""

    n = len(data)

    for pos in zip(*data):
        one_bit_count = pos.count("1")
        if one_bit_count >= n / 2:
            gamma += "1"
            epsilon += "0"
        else:
            gamma += "0"
            epsilon += "1"

    return int(gamma, 2) * int(epsilon, 2)


def main():
    import pathlib

    input_path = pathlib.Path(__file__).with_name("input.txt")

    with input_path.open() as f:
        print(compute(f.read().splitlines()))


if __name__ == "__main__":
    main()

File: day03/part1.cpp
#include <iostream>
#include <fstream>
#include <vector>


int compute(std::vector<std::string> numbers) {
    auto n = numbers.size();
    auto bitCount = numbers[0].size();

    auto gamma = 0;
    auto epsilon = 0;

    for (int i = 0; i < bitCount; ++i) {
        gamma <<= 1;
        epsilon <<= 1;

        int oneBitCount = 0;
        for (auto num : numbers) {
            if (num[i] == '1') {
                oneBitCount++;
            }
        }

        if (oneBitCount >= n - oneBitCount) {
            gamma += 1;
        } else {
            epsilon += 1;
        }
    }

    return gamma * epsilon;
}


int main() {
    std::string sourceFilePath = __FILE__;
    auto sourceDirPath = sourceFilePath.substr(0, sourceFilePath.rfind('\\'));
    auto inputPath = sourceDirPath + "/input.txt";

    std::ifstream file(inputPath);

    std::vector<std::string> numbers;

    if (!file) {
        std::cerr << "Couldn't open file " << inputPath << std::endl;
        return 1;
    }

    if (file.is_open()) {
        std::string line;
        while (std::getline(file, line)) {
            if (!line.empty()) {
                numbers.push_back(line);
            }
        }
        std::cout << compute(numbers) << std::endl;
        file.close();
    }

    return 0;
}
